File: smart_start_mobile/src/entities/offres/Projet.java
package entities.offres;

import java.util.Date;

public class Projet {
	
	
	private int id;
	private User client;
	private String nomProjet;
	private String TitreProjet;
	private Date DateDebut;
	private Date dateFin;
	private Date dateRealisation;
	private Date dateCreation;
	private Date datePublication;
	private String description;
	private boolean publie;
	private boolean valide;
	private boolean signale;
	private double cout;
	private int idClient;

	@Override
	public String toString() {
		return "Projet [id=" + id + ", client=" + client.getId() + ", nomProjet=" + nomProjet + ", TitreProjet=" + TitreProjet
				+ "]";
	}
	private boolean archive;
	
	public User getClient() {
		return client;
	}

	public void setClient(User client) {
		this.client = client;
	}
	
	public Projet(User client, String nomProjet, String titreProjet, Date dateDebut, Date dateFin,
			String description, boolean publie, boolean valide, boolean signale, double cout, boolean archive) {
		super();
		this.client = client;
		this.nomProjet = nomProjet;
		TitreProjet = titreProjet;
		DateDebut = dateDebut;
		this.dateFin = dateFin;
		this.description = description;
		this.publie = publie;
		this.valide = valide;
		this.signale = signale;
		this.cout = cout;
		this.archive = archive;
		this.idClient = client.getId();
	}

	public String getTitreProjet() {
		return TitreProjet;
	}


	public void setTitreProjet(String titreProjet) {
		TitreProjet = titreProjet;
	}



	public Date getDateDebut() {
		return DateDebut;
	}



	public void setDateDebut(Date dateDebut) {
		DateDebut = dateDebut;
	}



	public Date getDateFin() {
		return dateFin;
	}



	public void setDateFin(Date dateFin) {
		this.dateFin = dateFin;
	}



	public Date getDateRealisation() {
		return dateRealisation;
	}



	public void setDateRealisation(Date dateRealisation) {
		this.dateRealisation = dateRealisation;
	}



	public Date getDateCreation() {
		return dateCreation;
	}



	public void setDateCreation(Date dateCreation) {
		this.dateCreation = dateCreation;
	}



	public Date getDatePublication() {
		return datePublication;
	}



	public void setDatePublication(Date datePublication) {
		this.datePublication = datePublication;
	}



	public String getDescription() {
		return description;
	}



	public void setDescription(String description) {
		this.description = description;
	}



	public boolean isPublie() {
		return publie;
	}



	public void setPublie(boolean publie) {
		this.publie = publie;
	}



	public boolean isValide() {
		return valide;
	}



	public void setValide(boolean valide) {
		this.valide = valide;
	}



	public boolean isSignale() {
		return signale;
	}



	public void setSignale(boolean signale) {
		this.signale = signale;
	}



	public double getCout() {
		return cout;
	}



	public void setCout(double cout) {
		this.cout = cout;
	}



	public boolean isArchive() {
		return archive;
	}



	public void setArchive(boolean archive) {
		this.archive = archive;
	}

	
	
	public Projet( String nomProjet) {
		this.nomProjet = nomProjet;
	}



	public Projet() {
	}

	public int getId() {
		return id;
	}
	public void setId(int id) {
		this.id = id;
	}
	public String getNomProjet() {
		return nomProjet;
	}
	public void setNomProjet(String nomProjet) {
		this.nomProjet = nomProjet;
	}





}

File: smart_start_web/src/OffreBundle/Entity/Projet.php
<?php

namespace OffreBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Context\ExecutionContextInterface;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * Projet
 *
 * @ORM\Table(name="projet")
 * @ORM\Entity(repositoryClass="OffreBundle\Repository\ProjetRepository")
 */
class Projet
{
    /**
     * @var int
     *
     * @ORM\Column(name="id", type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     *
     * @var string
     *
     * @ORM\Column(name="nomProjet", type="string", length=255)
     */
    private $nomProjet;

    /**
     * @var string
     *
     * @ORM\Column(name="TitreProjet", type="string", length=255)
     */
    private $titreProjet;

    /**
     * @var \DateTime
     *
     * @ORM\Column(name="DateDebut", type="date", nullable=true)
     */
    private $dateDebut;

    /**
     * @var \DateTime
     * @Assert\GreaterThan(propertyPath="dateDebut")
     * @ORM\Column(name="dateFin", type="date", length=255, nullable=true)
     */
    private $dateFin;

    /**
     * @var \DateTime
     *
     * @ORM\Column(name="dateRealisation", type="date", nullable=true)
     */
    private $dateRealisation;

    /**
     * @var \DateTime
     *
     * @ORM\Column(name="dateCreation", type="date", nullable=true)
     */
    private $dateCreation;

    /**
     * @var \DateTime
     *
     * @ORM\Column(name="datePublication", type="date", nullable=true)
     */
    private $datePublication;

    /**
     * @var string
     *
     * @ORM\Column(name="description", type="text")
     */
    private $description;

    /**
     * @var bool
     *
     * @ORM\Column(name="publie", type="boolean", nullable=true)
     */
    private $publie;

    /**
     * @var bool
     *
     * @ORM\Column(name="valide", type="boolean")
     */
    private $valide;

    /**
     * @var bool
     *
     * @ORM\Column(name="signale", type="boolean", nullable=true)
     */
    private $signale;

    /**
     *
     * @ORM\OneToMany(targetEntity="freelancerParticipeProjet", mappedBy="projet")
     */
    private $participations;

    /**
     * Many features have one product. This is the owning side.
     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\User", inversedBy="$projetsClient"))
     * @ORM\JoinColumn(name="client_id", referencedColumnName="id")
     */
    private $client;

    /**
     * @var float
     *
     * @ORM\Column(name="cout", type="float")
     */
    private $cout;

    /**
     * @return float
     */
    public function getCout()
    {
        return $this->cout;
    }

    /**
     * @param float $cout
     */
    public function setCout($cout)
    {
        $this->cout = $cout;
    }


    /**
     * @return mixed
     */
    public function getClient()
    {
        return $this->client;
    }

    /**
     * @param mixed $client
     */
    public function setClient($client)
    {
        $this->client = $client;
    }


    /**
     * @return mixed
     */
    public function getDemandes()
    {
        return $this->demandes;
    }

    /**
     * @param mixed $demandes
     */
    public function setDemandes($demandes)
    {
        $this->demandes = $demandes;
    }


    /**
     *
     * @ORM\OneToMany(targetEntity="freelancerDemandeProjet", mappedBy="projet")
     */
    private $demandes;

    /**
     * @return mixed
     */
    public function getParticipations()
    {
        return $this->participations;
    }

    /**
     * @param mixed $participations
     */
    public function setParticipations($participations)
    {
        $this->participations = $participations;
    }

    /**
     * @return mixed
     */
    public function getTechnologies()
    {
        return $this->technologies;
    }

    /**
     * @param mixed $technologies
     */
    public function setTechnologies($technologies)
    {
        $this->technologies = $technologies;
    }


    /**
     * @ORM\ManyToMany(targetEntity="Technologie", inversedBy="projets")
     * @ORM\JoinTable(name="projet_technologie")
     */
    private $technologies;

    public function __construct() {
        $this->technologies = new \Doctrine\Common\Collections\ArrayCollection();
        $this->participations = new ArrayCollection();
    }

    public function addTechnologie(Technologie $technologie)
    {
        $technologie->addProjet($this); // synchronously updating inverse side
        $this->technologies[] = $technologie;
    }


    public function removeTechnologie($technologie)
    {
        return $this->technologies->removeElement($technologie);
    }

    public function removeDemande($demande)
    {
        return $this->demandes->removeElement($demande);
    }



    /**
     * Get id
     *
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }
    /**
     * Set nomProjet
     *
     * @param string $nomProjet
     *
     * @return Projet
     */
    public function setNomProjet($nomProjet)
    {
        $this->nomProjet = $nomProjet;
        return $this;
    }
    /**
     * Get nomProjet
     *
     * @return string
     */
    public function getNomProjet()
    {
        return $this->nomProjet;
    }

    /**
     * Set titreProjet
     *
     * @param string $titreProjet
     *
     * @return Projet
     */
    public function setTitreProjet($titreProjet)
    {
        $this->titreProjet = $titreProjet;

        return $this;
    }

    /**
     * Get titreProjet
     *
     * @return string
     */
    public function getTitreProjet()
    {
        return $this->titreProjet;
    }

    /**
     * Set dateDebut
     *
     * @param \DateTime $dateDebut
     *
     * @return Projet
     */
    public function setDateDebut($dateDebut)
    {
        $this->dateDebut = $dateDebut;

        return $this;
    }

    /**
     * Get dateDebut
     *
     * @return \DateTime
     */
    public function getDateDebut()
    {
        return $this->dateDebut;
    }

    /**
     * Set dateFin
     *
     * @param string $dateFin
     *
     * @return Projet
     */
    public function setDateFin($dateFin)
    {
        $this->dateFin = $dateFin;

        return $this;
    }

    /**
     * Get dateFin
     *
     * @return string
     */
    public function getDateFin()
    {
        return $this->dateFin;
    }

    /**
     * Set dateRealisation
     *
     * @param \DateTime $dateRealisation
     *
     * @return Projet
     */
    public function setDateRealisation($dateRealisation)
    {
        $this->dateRealisation = $dateRealisation;

        return $this;
    }

    /**
     * Get dateRealisation
     *
     * @return \DateTime
     */
    public function getDateRealisation()
    {
        return $this->dateRealisation;
    }

    /**
     * Set dateCreation
     *
     * @param \DateTime $dateCreation
     *
     * @return Projet
     */
    public function setDateCreation($dateCreation)
    {
        $this->dateCreation = $dateCreation;

        return $this;
    }

    /**
     * Get dateCreation
     *
     * @return \DateTime
     */
    public function getDateCreation()
    {
        return $this->dateCreation;
    }

    /**
     * Set datePublication
     *
     * @param \DateTime $datePublication
     *
     * @return Projet
     */
    public function setDatePublication($datePublication)
    {
        $this->datePublication = $datePublication;

        return $this;
    }

    /**
     * Get datePublication
     *
     * @return \DateTime
     */
    public function getDatePublication()
    {
        return $this->datePublication;
    }

    /**
     * Set description
     *
     * @param string $description
     *
     * @return Projet
     */
    public function setDescription($description)
    {
        $this->description = $description;

        return $this;
    }

    /**
     * Get description
     *
     * @return string
     */
    public function getDescription()
    {
        return $this->description;
    }

    /**
     * Set publie
     *
     * @param boolean $publie
     *
     * @return Projet
     */
    public function setPublie($publie)
    {
        $this->publie = $publie;

        return $this;
    }

    /**
     * Get publie
     *
     * @return bool
     */
    public function getPublie()
    {
        return $this->publie;
    }

    /**
     * Set valide
     *
     * @param boolean $valide
     *
     * @return Projet
     */
    public function setValide($valide)
    {
        $this->valide = $valide;

        return $this;
    }

    /**
     * Get valide
     *
     * @return bool
     */
    public function getValide()
    {
        return $this->valide;
    }

    /**
     * Set signale
     *
     * @param boolean $signale
     *
     * @return Projet
     */
    public function setSignale($signale)
    {
        $this->signale = $signale;

        return $this;
    }

    /**
     * Get signale
     *
     * @return bool
     */
    public function getSignale()
    {
        return $this->signale;
    }

    /**
     * @Assert\Callback
     */
    public function validate(ExecutionContextInterface $context, $payload)
    {
        if(strlen($this->nomProjet)<2){
            $context->buildViolation('Nom projet trop court')
                ->atPath('nomProjet')
                ->addViolation();
        }
        if(strlen($this->titreProjet)<2){
            $context->buildViolation('Titre du projet trop court')
                ->atPath('titreProjet')
                ->addViolation();
        }
        $jour_auj = date('j');
        $mois_auj = date('m');
        $annee_auj = date('y');
        if($this->dateDebut->format('y')-$annee_auj>=4 ){
            $context->buildViolation('Date de debut sup ou egale a 4')
                ->atPath('dateDebut')
                ->addViolation();
        }

        if(date_diff($this->dateFin,$this->dateDebut)->format('%R%a days') > 0 ){
            $context->buildViolation('La date de fin doit etre apres la date de debut')
                ->atPath('dateFin')
                ->addViolation();
        }
    }

}

